wdl-lint/src/rules/shellcheck.rs

```diff
@@ -0,0 +1,424 @@
+        Some(code) => bail!("unexpected `shellcheck` exit code: {}", code),
+        None => bail!("the `shellcheck` process appears to have been interrupted"),
+    }
+}
+
+/// Runs ShellCheck on a command section and reports diagnostics.
+#[derive(Default, Debug, Clone, Copy)]
+pub struct ShellCheckRule;
+
+impl Rule for ShellCheckRule {
+    fn id(&self) -> &'static str {
+        ID
+    }
+
+    fn description(&self) -> &'static str {
+        "Ensures that command blocks are free of ShellCheck violations."
+    }
+
+    fn explanation(&self) -> &'static str {
+        "ShellCheck (https://shellcheck.net) is a static analysis tool and linter for sh / bash. \
+         The lints provided by ShellCheck help prevent common errors and pitfalls in your scripts. \
+         Following its recommendations will increase the robustness of your command sections."
+    }
+
+    fn tags(&self) -> TagSet {
+        TagSet::new(&[Tag::Correctness, Tag::Portability])
+    }
+
+    fn exceptable_nodes(&self) -> Option<&'static [SyntaxKind]> {
+        Some(&[
+            SyntaxKind::VersionStatementNode,
+            SyntaxKind::CommandSectionNode,
+        ])
+    }
+}
+
+/// Convert a WDL `Placeholder` to a bash variable declaration.
+///
+/// Returns "WDL" + <placeholder length - 6> random alphnumeric characters.
+/// The returned value is shorter than the placeholder by 3 characters so
+/// that the caller may pad with other characters as necessary
+/// depending on whether or not the variable needs to be treated as a
+/// declaration, expansion, or literal.
+fn to_bash_var(placeholder: &Placeholder) -> String {
+    let placeholder_len: usize = placeholder.syntax().text_range().len().into();
+    // don't start variable with numbers
+    let mut bash_var = String::from("WDL");
+    bash_var.push_str(
+        &Alphanumeric.sample_string(&mut rand::thread_rng(), placeholder_len.saturating_sub(6)),
+    );
+    bash_var
+}
+
+/// Retrieve all input and private declarations for a task.
+fn gather_task_declarations(task: &TaskDefinition) -> HashSet<String> {
+    let mut decls = HashSet::new();
+    if let Some(input) = task.input() {
+        for decl in input.declarations() {
+            decls.insert(decl.name().as_str().to_owned());
+        }
+    }
+
+    for decl in task.declarations() {
+        decls.insert(decl.name().as_str().to_owned());
+    }
+    decls
+}
+
+/// Creates a "ShellCheck lint" diagnostic from a `ShellCheckDiagnostic`
+fn shellcheck_lint(diagnostic: &ShellCheckDiagnostic, span: Span) -> Diagnostic {
+    let label = format!(
+        "SC{}[{}]: {}",
+        diagnostic.code, diagnostic.level, diagnostic.message
+    );
+    Diagnostic::note(&diagnostic.message)
+        .with_rule(ID)
+        .with_label(label, span)
+        .with_label(
+            format!("more info: {}/SC{}", &SHELLCHECK_WIKI, diagnostic.code),
+            span,
+        )
+        .with_fix("address the diagnostic as recommended in the message")
```

Comment:
@a-frantz - I know you just added/updated our `fix` messages. What do you think of this one? The generic message isn't helpful to users, but the alternative would presumably involve parsing the warnings from shellcheck and writing a fix for each type.

Reply:
To be perfectly honest, yall didn't really convince me that it's the right move for every diagnostic to include a `fix`. I was just outvoted and didn't feel strongly enough to put up a fight 😅 

So my frank opinion is that this `fix` should be dropped _along with some of our other "unhelpful" fix messages_. But we've set the precedent that every diagnostic should include some kind of `fix`, and I'm not sure how we can improve this one without some rather intense work 🤷 

@thatRichman I see [here](https://github.com/stjude-rust-labs/wdl/pull/264/files#diff-9cabcff67f58680d419fec1e45eb2ac94d72365b16534b9351744b65ffa370f1R69) that `shellcheck` provides it's own `fix` message. Maybe we should just copy that? I haven't seen what they look like, so maybe they aren't of good quality 🤔 
Can you share some examples of the shellcheck fix messages?

Reply:
The fix messages describe the structural changes required to address the lint, and are not always available.
Example:
```
    "fix": {
      "replacements": [
        {
          "line": 4,
          "endLine": 4,
          "precedence": 7,
          "insertionPoint": "afterEnd",
          "column": 11,
          "replacement": "\"",
          "endColumn": 11
        },
        {
          "line": 4,
          "endLine": 4,
          "precedence": 7,
          "insertionPoint": "beforeStart",
          "column": 15,
          "replacement": "\"",
          "endColumn": 15
        }
      ]
```

We _could_ convert these into a human-readable message, but I can't say if we want to.

Reply:
@a-frantz - This could be used to get toward my goal for `fix`. I would want to use that `fix` to print the correct line. 
```
echo [ -z "$foo" ]
```

Reply:
Oo. That's kind of messy... At the very least it wouldn't be trivial to convert that to something I would want us to print.

@adthrasher I agree that we should work towards having every `fix` message print the "corrected" code where possible. But that's not very easy to implement robustly. This one in particular looks tough to parse _when it's present_.

My two cents on this issue is we leave the `fix` message as Spencer has written it, but we can make a new issue to revisit this specific `fix` (and many of our other fixes for that matter).

Reply:
@a-frantz - That particular example doesn't look too bad to convert to text. You've just got the line, column, replacement text, and how to place it. The bigger issue, to me, is that it isn't always present (though some errors probably aren't straightforward like this one). But I agree, we can leave it as-is and have a new issue to revisit.